libtransmission/settings.cc

```diff
@@ -3,15 +3,460 @@
 // or any future license endorsed by Mnemosyne LLC.
 // License text can be found in the licenses/ folder.
 
+#include <array>
+#include <cstddef> // size_t
+#include <cstdint> // int64_t, uint32_t
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
 #include <variant>
 
+#include <fmt/core.h>
+
+#include "libtransmission/transmission.h"
+
+#include "libtransmission/log.h" // for tr_log_level
+#include "libtransmission/net.h" // for tr_port
+#include "libtransmission/open-files.h" // for tr_open_files::Preallocation
+#include "libtransmission/peer-io.h" // tr_preferred_transport
 #include "libtransmission/settings.h"
+#include "libtransmission/utils.h" // for tr_strv_strip(), tr_strlower()
 #include "libtransmission/variant.h"
 
+using namespace std::literals;
+
 namespace libtransmission
 {
 namespace
 {
+struct VariantConverter
+{
+public:
+    template<typename T>
+    static std::optional<T> load(tr_variant const& src);
+
+    template<typename T>
+    static tr_variant save(T const& val);
+
+private:
+    template<typename T, size_t N>
+    using Lookup = std::array<std::pair<std::string_view, T>, N>;
+
+    static inline auto constexpr EncryptionKeys = Lookup<tr_encryption_mode, 3U>{ {
+        { "required", TR_ENCRYPTION_REQUIRED },
+        { "preferred", TR_ENCRYPTION_PREFERRED },
+        { "allowed", TR_CLEAR_PREFERRED },
+    } };
+
+    static inline auto constexpr LogKeys = Lookup<tr_log_level, 7U>{ {
+        { "critical", TR_LOG_CRITICAL },
+        { "debug", TR_LOG_DEBUG },
+        { "error", TR_LOG_ERROR },
+        { "info", TR_LOG_INFO },
+        { "off", TR_LOG_OFF },
+        { "trace", TR_LOG_TRACE },
+        { "warn", TR_LOG_WARN },
+    } };
+
+    static inline auto constexpr PreallocationKeys = Lookup<tr_open_files::Preallocation, 5U>{ {
+        { "off", tr_open_files::Preallocation::None },
+        { "none", tr_open_files::Preallocation::None },
+        { "fast", tr_open_files::Preallocation::Sparse },
+        { "sparse", tr_open_files::Preallocation::Sparse },
+        { "full", tr_open_files::Preallocation::Full },
+    } };
+
+    static inline auto constexpr VerifyModeKeys = Lookup<tr_verify_added_mode, 2U>{ {
+        { "fast", TR_VERIFY_ADDED_FAST },
+        { "full", TR_VERIFY_ADDED_FULL },
+    } };
+
+    static inline auto constexpr PreferredTransportKeys = Lookup<tr_preferred_transport, TR_NUM_PREFERRED_TRANSPORT>{ {
+        { "utp", TR_PREFER_UTP },
+        { "tcp", TR_PREFER_TCP },
+    } };
+};
+
+template<>
+std::optional<bool> VariantConverter::load<bool>(tr_variant const& src)
```

Comment:
`<T>` behind the function name can be omitted for all `save()` and `load()` specialisations.

Reply:
Good catch. Removed in c47f5f1891d1a853618d4b6f7dae811408fe3894